crates/ruff_python_semantic/src/statements.rs

```diff
@@ -0,0 +1,100 @@
+use std::ops::{Index, IndexMut};
+
+use ruff_index::{newtype_index, IndexVec};
+use ruff_python_ast::Stmt;
+use rustc_hash::FxHashMap;
+
+use ruff_python_ast::types::RefEquality;
+
+/// Id uniquely identifying a statement AST node.
+///
+/// Using a `u32` is sufficient because Ruff only supports parsing documents with a size of max
+/// `u32::max` and it is impossible to have more nodes than characters in the file. We use a
+/// `NonZeroU32` to take advantage of memory layout optimizations.
+#[newtype_index]
+#[derive(Ord, PartialOrd)]
+pub struct StatementId;
+
+/// A [`Stmt`] AST node, along with a pointer to its parent statement (if any).
+#[derive(Debug)]
+struct Statement<'a> {
+    /// A pointer to the AST node.
+    node: &'a Stmt,
+    /// The ID of the parent of this node, if any.
+    parent: Option<StatementId>,
+    /// The depth of this node in the tree.
+    depth: u32,
+}
+
+/// The statements of a program indexed by [`StatementId`]
+#[derive(Debug, Default)]
+pub struct Statements<'a> {
+    nodes: IndexVec<StatementId, Statement<'a>>,
+    node_to_id: FxHashMap<RefEquality<'a, Stmt>, StatementId>,
+}
+
+impl<'a> Statements<'a> {
+    /// Inserts a new statement into the node tree and returns its unique ID.
+    ///
+    /// Panics if a statement with the same pointer already exists.
+    pub(crate) fn insert(&mut self, node: &'a Stmt, parent: Option<StatementId>) -> StatementId {
+        let next_id = self.nodes.next_index();
+        if let Some(existing_id) = self.node_to_id.insert(RefEquality(node), next_id) {
+            panic!("Node already exists with id {existing_id:?}");
+        }
+        self.nodes.push(Statement {
+            node,
+            parent,
+            depth: parent.map_or(0, |parent| self.nodes[parent].depth + 1),
+        })
+    }
+
+    /// Returns the [`StatementId`] of the given statement.
+    #[inline]
+    pub fn node_id(&self, statement: &'a Stmt) -> Option<StatementId> {
+        self.node_to_id.get(&RefEquality(statement)).copied()
+    }
+
+    /// Return the [`StatementId`] of the parent statement.
+    #[inline]
+    pub fn parent_id(&self, statement_id: StatementId) -> Option<StatementId> {
+        self.nodes[statement_id].parent
+    }
+
+    /// Return the parent of the given statement.
+    pub fn parent(&self, statement: &'a Stmt) -> Option<&'a Stmt> {
```

Comment:
Nit: The `parent` and `parent_id` methods are asymetric in that one accepts a `StatementId` and the other `Stmt`. Are we back to `statement_parent` 😆 ? Or `resolve_parent` or require two steps: `statement_id(stmt)` and then call `parent`?

Reply:
Haha. I will fix this in a follow-up since this API already exists on `Nodes` (now removed).